pkg/virt-controller/watch/replicaset.go

```diff
@@ -0,0 +1,533 @@
+	filtered := []virtv1.VirtualMachine{}
+	for _, vm := range vms {
+		if !vm.IsFinal() {
+			filtered = append(filtered, vm)
+		}
+	}
+	return filtered
+}
+
+// filterMatchingVMs takes a selector and a list of VMs. If the VM labels match the selector it is added to the filtered collection.
+// Returns the list of all VMs which match the selector
+func (c *VMReplicaSet) filterMatchingVMs(selector labels.Selector, vms []virtv1.VirtualMachine) []virtv1.VirtualMachine {
+	//TODO take owner reference into account
+	filtered := []virtv1.VirtualMachine{}
+	for _, vm := range vms {
+		if selector.Matches(labels.Set(vm.ObjectMeta.Labels)) {
+			filtered = append(filtered, vm)
+		}
+	}
+	return filtered
+}
+
+// listVMsFromNamespace takes a namespace and returns all VMs from the VM cache which run in this namespace
+func (c *VMReplicaSet) listVMsFromNamespace(namespace string) ([]virtv1.VirtualMachine, error) {
+	objs, err := c.vmInformer.GetIndexer().ByIndex(cache.NamespaceIndex, namespace)
+	if err != nil {
+		return nil, err
+	}
+	vms := []virtv1.VirtualMachine{}
+	for _, obj := range objs {
+		vms = append(vms, *obj.(*virtv1.VirtualMachine))
+	}
+	return vms, nil
+}
+
+// listControllerFromNamespace takes a namespace and returns all VMReplicaSets from the ReplicaSet cache which run in this namespace
+func (c *VMReplicaSet) listControllerFromNamespace(namespace string) ([]virtv1.VirtualMachineReplicaSet, error) {
+	objs, err := c.vmRSInformer.GetIndexer().ByIndex(cache.NamespaceIndex, namespace)
+	if err != nil {
+		return nil, err
+	}
+	replicaSets := []virtv1.VirtualMachineReplicaSet{}
+	for _, obj := range objs {
+		rs := obj.(*virtv1.VirtualMachineReplicaSet)
+		replicaSets = append(replicaSets, *rs)
+	}
+	return replicaSets, nil
+}
+
+// getMatchingController returns the first VMReplicaSet which matches the labels of the VM from the listener cache.
+// If there are no matching controllers, a NotFound error is returned.
+func (c *VMReplicaSet) getMatchingController(vm *virtv1.VirtualMachine) (*virtv1.VirtualMachineReplicaSet, error) {
+	log := logging.DefaultLogger()
+	controllers, err := c.listControllerFromNamespace(vm.ObjectMeta.Namespace)
+	if err != nil {
+		return nil, err
+	}
+
+	// TODO check owner reference, if we have an existing controller which owns this one
+
+	for _, rs := range controllers {
+		selector, err := v1.LabelSelectorAsSelector(rs.Spec.Selector)
+		if err != nil {
+			log.Error().Object(&rs).Reason(err).Msg("Failed to parse label selector from replicaset.")
+			continue
+		}
+
+		if selector.Matches(labels.Set(vm.ObjectMeta.Labels)) {
+			// The first matching rs will be returned
+			return &rs, nil
+		}
+
+	}
+	return nil, errors.NewNotFound(v1beta1.Resource("virtualmachinereplicaset"), "")
+}
+
+// addVirtualMachine searchs for a matching VMReplicaSet, updates it's expectations and wakes it up
+func (c *VMReplicaSet) addVirtualMachine(obj interface{}) {
```

Comment:
add/del/update functions are all doing the same thing minus 1 line involving expectations.  Maybe create a helper function?

Reply:
:+1: